### ISSUE TITLE
Booleans aren't sufficiently Boolean-y

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#697]
**Describe the bug**
Given the following setup:

1) In one twinpack project we have:
```
'int sqlite3_compileoption_used(const char *zOptName);
Public DeclareWide PtrSafe Function sqlite3_compileoption_used _
CDecl Lib "sqlite3.dll" ( _
  ByVal zOptName As LongPtr _
) As Boolean	
```

2) In the other project referencing the prior project, we have:
```
Public Function CompileOptionUsed(OptionName As String) As Boolean
	If Not IsLoaded Then
		Err.Raise 5, CurrentComponentName, FormatErrorMessage(CurrentProcedureName)
	End If
	Return Sqlite3.sqlite3_compileoption_used(StrPtr(StrConv(OptionName, vbFromUnicode)))
End Function
```

3) We then have the following test method:
```
[ TestCase ]
Private Sub Test_CompileOptionUsed()
	If (Sqlite3Instance.Load(DllPath) = False) Then
		Assert.Strict.Inconclusive
	End If
		
	Dim Option As String
	Option = CompileOptionGet(0)
	Assert.Strict.IsTrue(CompileOptionUsed(Option))
End Sub
```

We get an assertion failed even though the result is "True". See screenshot and also note the Debug Console testing different ways:
![image](https://user-images.githubusercontent.com/2367644/151663476-0b90b059-4526-4bef-9f3b-0d9fa896cb0f.png)

I went into settings and changed the option to sanitize Booleans from default `No` to `Yes`:
![image](https://user-images.githubusercontent.com/2367644/151663498-4ff486fd-b4fa-41d9-8e1e-7301b8ecb150.png)

Running the test code then yields a different error:
![image](https://user-images.githubusercontent.com/2367644/151663532-215f8095-0b90-4a77-a1d8-6ba8882099c6.png)

I then go back to the first project, set the same Sanitize Boolean property, publish a new version of the twinpack. I return to the 2nd project, update the reference. I still get the same codegen error. So I revert the Sanitize Boolean setting back to its default (`No`) in the 2nd project. Despite the first project still having "Sanitize Boolean" property set to "yes", I then get the same assertion failed error in the 2nd project so it did not seem to do anything. 

**Expected behavior**
[According to this page](https://www.sqlite.org/c3ref/compileoption_get.html) the API should return either 0 or 1. Being a C export, there is of course no native Boolean data type. So we'd get a "dirty" Boolean which won't equal to `True`. So while it is expected that the `Assert.Strict.IsTrue` should fail, I'd have expected that either:

* `Assert.Permissive.IsTrue` would succeed even with a dirty Boolean
* `Sanitize Boolean` setting would work across twinpack boundary and not result in a codegen error.

NB: I can work around the dirty Boolean with this:
```
If CompileOptionUsed(Option) Then
	Assert.Strict.Succeed
Else
	Assert.Strict.Fail
End If
```

**Desktop (please complete the following information):**
 - OS: Windows 7
 - twinBASIC compiler version 0.13.91

**Comment by WaynePhillipsEA** [comment]
Yes, I think the permissive assert should allow for that case, and a strict/exact assert should fail (but should also indicate that the True is dirty in the actual/expected printed value).

I'll also check the Sanitize-option codegen, as clearly something is not right there.

**Comment by Kr00l** [comment]
`int sqlite3_compileoption_used`

Isn't Boolean wrong in your Declare?
`int` is 32 bit (As Long) but Boolean is 16 bit.

**Comment by bclothier** [comment]
My understanding is that it's safe to use `Boolean` since VBx will automatically handle the conversion. I think it's more semantically correct to convert `BOOL` to `Boolean` rather than ` Long` (which is ambiguous -- is it a Boolean result or is it a function that returns a number? )

**Comment by WaynePhillipsEA** [comment]
Provided it doesn't return something only in the upper half of the DWORD for meaning True e.g. (&H12340000), then Boolean will be fine with the Sanitize-option on.  Usually the return value will be TRUE(1)/FALSE(0) from these types of APIs, so in those cases it would be fine to use Boolean with Sanitize-option ON... when the codegen error is fixed.

**Comment by Kr00l** [comment]
But doesn't it cause a memory risk? The declare reads 2 bytes but the API lib writes 4 bytes even if just the LoDword is meaningful. (0 or 1)

**Comment by WaynePhillipsEA** [comment]
No, return values here are passed in a register (EAX/RAX), so it's technically passed as 32-bit regardless (or 64-bit on x64).

**Comment by bclothier** [comment]
FWIW:

Given the C source code:
```
int __declspec(dllexport) test() {
	return 0xFFFF0000;
}
```

and the declare from 64-bit Excel:
```
Public Declare PtrSafe Function Test_Long Lib "C:\...\TestCExport.dll" Alias "test" () As Long
Public Declare PtrSafe Function Test_Boolean Lib "C:\...\TestCExport.dll" Alias "test" () As Boolean
```

We get:
```
?Test_Long
-65536 

?Test_Boolean
False
```

I've never actually seen a "boolean" C export that would not return something in the lower portion of the word. Would make for a very strange API if it did. Is there a legitimate reason to have a "boolean" result that's within the higher portion of the word only?

**Comment by bclothier** [comment]
> No, return values here are passed in a register (EAX/RAX), so it's technically passed as 32-bit regardless (or 64-bit on x64).

and what about the parameters? I can't recall offhand but I'd think it also works even if it's off the stack. At least I do not remember crashing on a `Boolean` as a parameter.

**Comment by WaynePhillipsEA** [comment]
> Is there a legitimate reason to have a "boolean" result that's within the higher portion of the word only?

If it's declared as BOOL, the API should only ever be returning TRUE/FALSE.  

>and what about the parameters? I can't recall offhand but I'd think it also works even if it's off the stack. At least I do not remember crashing on a Boolean as a parameter.

The stack is aligned to 32/64 bit boundaries, so it is safe for *ByVal* arguments.  It is most definitely NOT safe for *ByRef* arguments though.

**Comment by bclothier** [comment]
Answered my own question (too slowly):

C:
```
int __declspec(dllexport) test(int bool) {
	return bool;
}
```

64-bit Excel:
```
Public Declare PtrSafe Function Test_Long Lib "C:\...\TestCExport.dll" Alias "test" (ByVal i As Long) As Long
Public Declare PtrSafe Function Test_Boolean Lib "C:\...\TestCExport.dll" Alias "test" (ByVal i As Boolean) As Boolean
```

Output:
```
?Test_Long(0)
 0 
?Test_Long(1)
 1 
?Test_Long(2)
 2 
?Test_Boolean(False)
False
?Test_Boolean(0)
False
?Test_Boolean(True)
True
?Test_Boolean(1)
True
?Test_Boolean(1) = True
True
```

> It is most definitely NOT safe for ByRef arguments though.
I think that's the problem - I don't remember seeing API that passes boolean by reference. 

However, new test...

C:
```
int __declspec(dllexport) test(int* bool) {
	return *bool;
}
```

Excel 64-bit:
```
Public Declare PtrSafe Function Test_Long Lib "C:\...\TestCExport.dll" Alias "test" (ByRef i As Long) As Long
Public Declare PtrSafe Function Test_Boolean Lib "C:\...\TestCExport.dll" Alias "test" (ByRef i As Boolean) As Boolean
Public Declare PtrSafe Function Test_Boolean2 Lib "C:\...\TestCExport.dll" Alias "test" (ByRef i As Long) As Boolean
```

Test:
```
i&=0:?Test_Long(i)
 0 
i&=1:?Test_Long(i)
 1 
i&=2:?Test_Long(i)
 2 
?Test_Boolean(False)
False
?Test_Boolean(True)
True
?Test_Boolean2(0)
False
?Test_Boolean2(-1)
True
?Test_Boolean2(&HFFFF0000&)
False
i&=&HFFFF0000&:?Test_Boolean2(i)
False
```

No crash...

**Comment by WaynePhillipsEA** [comment]
> No crash...

In your `test(int* bool)` implementation, you're not writing out to that argument.  If you did, you'd be overwriting the stack adjacent to the variable where you pass in a pointer to a Boolean on the VBA side, which will likely be another variable... or you may get lucky and it just be padding.   Ultimately, it's not safe, and you'll be corrupting memory.

**Comment by Kr00l** [comment]
I once had a nasty bug with SendMessage where lParam is As Any in the declare.
I passed `ByVal IntVariable` and the other side expected a Long variable. (I wasn't aware I have a Integer just..)
Most of the time it just worked as the missing HiWord was maybe 0, but sometimes it readed something corrupt. It took a while to realize this  and the fix was easy to make ByVal CLng(IntVariable)

That's why I'm hyper correct to match everything exactly.

**Comment by bclothier** [comment]
The unfortunate irony is that the advice to use `Boolean` was based on a very old VB6 site I once read years and years ago and the guy claimed that VB6 handles the conversion automatically. I was long aware of the dirty booleans, which is still an issue even without API (there are type libraries or even pure VBA code that can create dirty booleans). I do not think I ever encountered a win32 API that passed "booleans" by reference which would be why I never had a problem with using `Boolean`s for `BOOL` data types from win32 API.

**Comment by bclothier** [comment]
Just to confirm Wayne's remark:
C:
```
int __declspec(dllexport) test(int* bool) {
	int tmp = *bool;
	*bool = 0xFFFF0000;
	return tmp;
}
```

64-bit Excel:
```
Public Declare PtrSafe Function Test_Long Lib "C:\...\TestCExport.dll" Alias "test" (ByRef i As Long) As Long
Public Declare PtrSafe Function Test_Boolean Lib "C:\...\TestCExport.dll" Alias "test" (ByRef i As Boolean) As Boolean
Public Declare PtrSafe Function Test_Boolean2 Lib "C:\Users\BenClothier\Source\Repos\TestCExport\x64\Debug\TestCExport.dll" Alias "test" (ByRef i As Long) As Boolean

Public Type T
    i1 As Integer
    b As Boolean
    i2 As Integer
End Type
Public T As T
```

Test:
```
t.i1=1:t.b=True:t.i2=2:?t.i1,t.b,t.i2:?Test_Boolean(t.b):?t.i1,t.b,t.i2
 1            True           2 
True
 1            False         -1 
```

Not pretty!

**Comment by bclothier** [comment]
An additional thought. In .NET, we can control the marshaling behavior with the [`MarshalAs` attribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshalasattribute?view=net-6.0). There, we would probably do something like:

```
[return: MarshalAs(UnmanagedType.I4)]
bool test([MarshalAs(UnmanagedType.I4)] ref bool i)
```

This way, we have more control over the semantics of external API and would be more flexible than a global Sanitize-Boolean option as I'm sure there are other types that needs to be handled differently without sacrificing the semantics.

**Comment by bclothier** [comment]
> An additional thought. ...

Thinking about it some more, I'm actually inclined toward implementing the proposal in #619 which would do a much better job at describing the semantics while allowing us to control the conversion without relying on a magical attribute. In future, some of the work could be abstracted out into attributes to reduce the boilerplate if we think it merits it but that's an entirely separate discussion anyway.
